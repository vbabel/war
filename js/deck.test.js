const Deck = require('./deck');

describe('constructor', () => {
  test('builds a deck with the max number of suits and ranks by default', () => {
    const deck = new Deck();

    expect(deck.cards).toMatchSnapshot();
  });

  test('builds a deck with the supplied number of suits and ranks', () => {
    const deck = new Deck(2, 6);

    expect(deck.cards.length).toBe(12);
    expect(deck.cards).toMatchSnapshot();
  });

  test('throws an error when 0 suits are passed', () => {
    expect(() => new Deck(0)).toThrow(Deck.ERRORS.INVALID_NUM_SUITS);
  });

  test('throws an error when more than the max number of suits are passed', () => {
    expect(() => new Deck(Deck.MAX_SUITS + 1)).toThrow(Deck.ERRORS.INVALID_NUM_SUITS);
  });

  test('throws an error when 0 ranks are passed', () => {
    expect(() => new Deck(1, 0)).toThrow(Deck.ERRORS.INVALID_NUM_RANKS);
  });

  test('throws an error when more than the max number of ranks are passed', () => {
    expect(() => new Deck(1, Deck.MAX_RANKS + 1)).toThrow(Deck.ERRORS.INVALID_NUM_RANKS);
  });
});

describe('shuffle', () => {
  //This test has a very slight probability of failing
  test('shuffles the deck', () => {
    const deck = new Deck();
    const oldCards = deck.cards.slice(0);

    deck.shuffle();
    expect(deck.cards).not.toEqual(oldCards);
  });
});

describe('deal', () => {
  test('pops and returns the top card of the deck', () => {
    const deck = new Deck();
    deck.shuffle();

    const topCard = deck.cards[deck.cards.length - 1];

    expect(deck.deal()).toEqual(topCard);
  });

  test('returns undefined when the deck is empty', () => {
    const deck = new Deck(1, 1);
    deck.deal();

    expect(deck.deal()).toBeUndefined();
  });
});
