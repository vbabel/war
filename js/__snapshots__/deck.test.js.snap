// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor builds a deck with the max number of suits and ranks by default 1`] = `
Array [
  Card {
    "rank": Object {
      "name": "Two",
      "shortName": "2",
      "value": 2,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Three",
      "shortName": "3",
      "value": 3,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Four",
      "shortName": "4",
      "value": 4,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Five",
      "shortName": "5",
      "value": 5,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Six",
      "shortName": "6",
      "value": 6,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Seven",
      "shortName": "7",
      "value": 7,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Eight",
      "shortName": "8",
      "value": 8,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Nine",
      "shortName": "9",
      "value": 9,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Ten",
      "shortName": "10",
      "value": 10,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Jack",
      "shortName": "J",
      "value": 11,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Queen",
      "shortName": "Q",
      "value": 12,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "King",
      "shortName": "K",
      "value": 13,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Ace",
      "shortName": "A",
      "value": 14,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Two",
      "shortName": "2",
      "value": 2,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Three",
      "shortName": "3",
      "value": 3,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Four",
      "shortName": "4",
      "value": 4,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Five",
      "shortName": "5",
      "value": 5,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Six",
      "shortName": "6",
      "value": 6,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Seven",
      "shortName": "7",
      "value": 7,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Eight",
      "shortName": "8",
      "value": 8,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Nine",
      "shortName": "9",
      "value": 9,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Ten",
      "shortName": "10",
      "value": 10,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Jack",
      "shortName": "J",
      "value": 11,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Queen",
      "shortName": "Q",
      "value": 12,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "King",
      "shortName": "K",
      "value": 13,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Ace",
      "shortName": "A",
      "value": 14,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Two",
      "shortName": "2",
      "value": 2,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Three",
      "shortName": "3",
      "value": 3,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Four",
      "shortName": "4",
      "value": 4,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Five",
      "shortName": "5",
      "value": 5,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Six",
      "shortName": "6",
      "value": 6,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Seven",
      "shortName": "7",
      "value": 7,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Eight",
      "shortName": "8",
      "value": 8,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Nine",
      "shortName": "9",
      "value": 9,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Ten",
      "shortName": "10",
      "value": 10,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Jack",
      "shortName": "J",
      "value": 11,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Queen",
      "shortName": "Q",
      "value": 12,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "King",
      "shortName": "K",
      "value": 13,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Ace",
      "shortName": "A",
      "value": 14,
    },
    "suit": Object {
      "name": "Hearts",
      "shortName": "♡",
    },
  },
  Card {
    "rank": Object {
      "name": "Two",
      "shortName": "2",
      "value": 2,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Three",
      "shortName": "3",
      "value": 3,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Four",
      "shortName": "4",
      "value": 4,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Five",
      "shortName": "5",
      "value": 5,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Six",
      "shortName": "6",
      "value": 6,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Seven",
      "shortName": "7",
      "value": 7,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Eight",
      "shortName": "8",
      "value": 8,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Nine",
      "shortName": "9",
      "value": 9,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Ten",
      "shortName": "10",
      "value": 10,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Jack",
      "shortName": "J",
      "value": 11,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Queen",
      "shortName": "Q",
      "value": 12,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "King",
      "shortName": "K",
      "value": 13,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
  Card {
    "rank": Object {
      "name": "Ace",
      "shortName": "A",
      "value": 14,
    },
    "suit": Object {
      "name": "Spades",
      "shortName": "♤",
    },
  },
]
`;

exports[`constructor builds a deck with the supplied number of suits and ranks 1`] = `
Array [
  Card {
    "rank": Object {
      "name": "Two",
      "shortName": "2",
      "value": 2,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Three",
      "shortName": "3",
      "value": 3,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Four",
      "shortName": "4",
      "value": 4,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Five",
      "shortName": "5",
      "value": 5,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Six",
      "shortName": "6",
      "value": 6,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Seven",
      "shortName": "7",
      "value": 7,
    },
    "suit": Object {
      "name": "Clubs",
      "shortName": "♧",
    },
  },
  Card {
    "rank": Object {
      "name": "Two",
      "shortName": "2",
      "value": 2,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Three",
      "shortName": "3",
      "value": 3,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Four",
      "shortName": "4",
      "value": 4,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Five",
      "shortName": "5",
      "value": 5,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Six",
      "shortName": "6",
      "value": 6,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
  Card {
    "rank": Object {
      "name": "Seven",
      "shortName": "7",
      "value": 7,
    },
    "suit": Object {
      "name": "Diamonds",
      "shortName": "♢",
    },
  },
]
`;
